Adds new page about human readable addresses

GBKS opened the pull request:
This is something we've discussed for a few weeks in the [this issue](https://github.com/BitcoinDesign/Meta/issues/638) and the UX channel in Discord and turns the findings into a page.

Primary focus of the page is on the [DNS-based payment information proposal](https://github.com/bitcoin/bips/pull/1551), with an additional paragraph about Lightning Address and a small note about PayNyms.

The UI mock-ups are more directional than implementation-ready, which I think is fine. But we could flesh this out further in a revision at a later point.

I think there's more work to do in cross-linking this from all the right places. I updated a few spots, but there's probably more. The page is at a point where more review and public feedback would be helpful in identifying missing and incorrect bits.

**If you have criticism of the [DNS-based payment information proposal](https://github.com/bitcoin/bips/pull/1551), please post it in that PR, and not here. This page is a reflection of the logic described in that PR.**

🧐[Check the preview](https://deploy-preview-1082--bitcoin-design-site.netlify.app/guide/how-it-works/human-readable-addresses/)🧐

johnsBeharry commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,273 @@
+# Human readable addresses
+
+
+[Bitcoin addresses]({{'/guide/glossary/address/' | relative_url}}) and [payment requests]({{'/guide/how-it-works/payment-request-formats/' | relative_url}}) are long and cryptic, and practically impossible to read, memorize, pronounce, understand, and type. Creating a separate format for human-readable addresses is a logical solution to this problem, and there have been multiple attempts at makin ...
```
```suggestion
[Bitcoin addresses]({{'/guide/glossary/address/' | relative_url}}) and [payment requests]({{'/guide/how-it-works/payment-request-formats/' | relative_url}}) are long and cryptic, practically impossible to read, memorize, pronounce, understand, or type. We are used to being able to communicate our social media handles, or email addresses easily. That expectation has also extended to transacting value, with neo-banking and payment applications. There have been multiple attempts at making human-readable addresses a reality in bitcoin.
```

johnsBeharry commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,273 @@
+- Familiarity: A format that feels more human and familiar to users
+- Ease of sharing: A format that is easier to remember and recall and share in a social setting
+
+Note that human-readable addresses should be <b>only</b> used for direct communication. Whenever payment information is communicated digitally (copy & paste, links, QR codes, messaging, etc), regular addresses should be used for best privacy and security.
```
This would be impossible to enforce on any level and is also not how it is currently implemented. See Nostr.

GBKS replied:
Sure thing. Just like address re-use, wallets can encourage best practices, but can't enforce them. That's why I wrote `should`. Or am I misunderstanding?

johnsBeharry replied:
mm gotcha - well then maybe then bolden the should and add "ideally".

GBKS commented:
Thank you for the super-fast feedback.

> Use of ".btc" as an example, which is a Stacks-oriented tld. Stacks is kind of a scam, and while non-ICANN domains are totally possible, this adds some nuanced confusion, particularly in that there can be many different namespaces with overlapping TLDs.

Thanks for pointing that out. I just pushed a commit updating the sample domains. I had checked whether there are official .btc domains, but was not aware of this "other" use.

> the B prefix is not practical or necessary.

There was a lot of discussion around the `₿` prefix, and I also asked various people directly. From what I heard, the opinions seem to be really split with no clear majority/consensus. Personally, I agree with the trust argument. The prefix makes it 100% clear that it's a bitcoin address and nothing else. For a casual user, who is not aware of the technical realities of what can and can't go wrong when mixing up emails and bitcoin addresses, this might be an important signal. But I also understand the other way of approaching this argument. Is the better place to discuss this in the original proposal, which this page follows?

Also, did you just mention using a sat symbol? 😉

> While I have no special care for/against Paynyms, it seems odd that the text here is so dismissive about it and centralization when all of the other patterns on this page require it too.

True, it is pretty dismissive. But PayNyms is basically a single database, while the other proposals do their best to decentralize, but are bound by the "basic infrastructure of the internet". Ver different ends of the spectrum to me. But if it is too dismissive, then it should be revised.

> While it is interesting to show UX for helping users set up their DNS TXT entry and their own self-hosted payment address, this is a very unlikely pattern, particularly in mobile. Most people wouldn't do this and use a trusted provider, and such likelihood and risk are important to focus on.

Agree, this needs more fleshing out. I mention in the page that the UI mocks are conceptual, and I'd like to take more time to discuss and sort this out. I still have some open questions, but also wanted to get this page out.

How do you think this will play out? Dedicated bitcoin address services like we have in Nostr for NIP-05 identifiers?

Thanks again for the feedback.

BitcoinErrorLog commented:
With Paynyms, I'm a bit too ignorant, but I'm skeptical they are any more centralized than DNS. All namespaces require an authoritative list to exist somewhere. And anyone can host a new list, or query any existing list they want to. Centralized/decentralized isn't really the correct paradigm for namespaces. 

Regarding how I think this will all play out, well, I think it is all moot tbh, and that such things are a fad born of ignorance. No one needs nicknames at a protocol level, they are app-level concerns and thus should be rendered in contact list UX. Domain names are a scam, and thus pay-names follow in kind. This is my, likely unpopular, opinion, so it could also be wrong :)

Think of DNS more like a list of phone numbers. I can either make a list myself, or trust someone else to provide a list (like a phone book/directory), but the difference is all about there being a trusted authority, thus solutions that do not require such an authority are more interesting, or, solutions that aggregate many lists, etc. 

If a user wants to pay "Mom" they should be able to just tap on Mom's face, not type in her fake email address into some novel UX.

johnsBeharry commented:
[PayNyms](http://paynym.is) are perhaps a bit more centralised than a DNS based option solely cause more people host the DNS to HTTP resolvers ([1](https://developers.cloudflare.com/1.1.1.1/encryption/dns-over-https/), [2](https://developers.google.com/speed/public-dns/docs/doh), [3](https://support.mozilla.org/en-US/kb/firefox-dns-over-https)), and such resolvers are open source. With PayNyms only samurai maintain a registry and doesn't look like they have open sourced it.

johnsBeharry commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,273 @@
+
+Note that human-readable addresses should be <b>only</b> used for direct communication. Whenever payment information is communicated digitally (copy & paste, links, QR codes, messaging, etc), regular addresses should be used for best privacy and security.
+
+On this page, we focus primarily on the [DNS Payment Instructions proposal](https://github.com/bitcoin/bips/pull/1551), with some information about [Lightning Addresses](https://lightningaddress.com). They pursue similar goals, but in very different ways.
```
We have promoted things which do not have wide adoption as yet like [Sign in with bitcoin](https://bitcoin.design/guide/how-it-works/sign-in-with-bitcoin) but here it's spoken about in the abstract instead of strong recommendation to WebLN over LNURL-Auth for example.

From my recollection this is the first time a payment protocol will be added to the guide at such an early stage, against something else which has already had adoption that it's to supplant.

Stephen created a separate campaign page for Unified QR Codes to help drive its' adoption. Will this page be the primary campaign page for the BIP?

Asking because we have been quite reserved so far, so wondering if it's a shift in our editorial policy for BDG.

GBKS replied:
Good point. I'll add that this is an early proposal.

Overall, I think we can be more forward-looking, especially if it could maybe help the ecosystem figure out how to move forward. We just need the proper disclaimers.

BitcoinErrorLog replied:
Please note that all replies here are hinting that the forward-looking aspects are the most questionable/doubtful.

GBKS replied:
Also see the [Stabilizing bitcoin value](https://bitcoin.design/guide/how-it-works/stabilizing-bitcoin-value/) which has a disclaimer about the techniques being experimental and such.

> Please note that all replies here are hinting that the forward-looking aspects are the most questionable/doubtful.

Isn't it normal and to be expected that the forward-looking aspects would be the ones that are discussed the most?

> Will this page be the primary campaign page for the BIP?

I do not know how this page will be used. There are no plans for a campaign. Matt made a proposal, this page is about how this proposal could surface in UIs.

GBKS replied:
@johnsBeharry I added a note in the page about that proposal being brand new and still being discussed.

johnsBeharry commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,273 @@
+
+## [Paynyms](http://paynym.is)
+
+This approach relies on a centralized directory/server and [BIP-47](https://github.com/bitcoin/bips/blob/master/bip-0047.mediawiki) payment codes. Centralization allows for the omission of the global part and users can simply be referred to by their usernames. However, it is not recommended due to the problematic centralization.
```
```suggestion
This approach relies on a single directory provider which maps a human readable name with a [BIP-47](https://github.com/bitcoin/bips/blob/master/bip-0047.mediawiki) payment code. Having a single provider allows for the omission of the global part, in this case "my.paynym.is/username" and users can simply be referred to by their usernames. However, it is not recommended due to the problematic centralization.
```

sbddesign commented:
I don't see the value in using the B symbol (which I frankly don't even know how to make with my keyboard). Can somebody educate me on the dangers of mixing up an email address and a bitcoin payment identifier?

alltheseas commented:
The "B" symbol is a recipe for frustration, and setting up normies for failure.

> but users shouldn't need to type it.

Why limit the design space of what users can, and can't do? 

> The reason its there is because payment instructions are a different namespace from email. We've already seen at least one case where a company uses their domain for emails and also to provide users with lnaddresses, with someone having the lnaddress who doesn't work for the company and doesn't have the corresponding email.

That's fine - add some other way instead of an impossible to type "B" symbol. 

There's "human readable", and there should be, in my view, "human typable". These are the most basic human machine interface requirements. 

Normies will not memorize the special character shortcut for "B". We don't have "B" on keyboards today. 


Really, anything else on the keyboard, or perhaps even on the more difficult to access emoji keyboard reduces user friction. The designers I'm sure can come up with something better.

Examples: BTC.name@domain
SATSname@
ZAPname@
LNname
🌽name@ (or any other emoji)

GBKS commented:
> The "B" symbol is a recipe for frustration, and setting up normies for failure.

To keep things a bit orderly, I'd like to ask for criticism of the proposal to be posted on the [original proposal pull request](https://github.com/bitcoin/bips/pull/1551). This new page in the guide is a reflection of the logic in that proposal. I'll also add this note to the PR description. Proposal criticism is not really something that can be resolved in comments here.

> There's "human readable", and there should be, in my view, "human typable". These are the most basic human machine interface requirements.

Totally. The page mentions "read, memorize, pronounce, understand, and type".

sbddesign commented on assets/images/guide/how-it-works/human-readable-addresses/update@2x.png:
While I personally like the idea of just hanging on to the offer, this seems to contradict [this text](https://github.com/bitcoin/bips/pull/1551/files) from the BIP.

> While clients MAY cache the payment instructions they receive from the DNS, clients MUST NOT cache the payment instructions received from the DNS for longer than the TTL provided by their DNS resolver, and further MUST NOT cache the payment instructions for longer than the lowest initial TTL (which is signed as a part of DNSSEC signatures) received in the full DNSSEC chain leading from the DNS root to the resolved TXT record.

So on a technical level, the client (wallet) can't cache the pay instructions for longer than the TTL. So it means it needs to recheck if the TTL is elapsed. In practice, I don't think you'd actually have the client re-checking every single person in the contact list everyday, just re-checking at time of payment.

Furthermore, I'm not sure how actionable it is -- you hit view changes, and then it shows you that it changed from one nasty looking `lnobc1....` string to another nasty `lnobc1...` string.

Is this going to notify me every single time the DNS record changes? Or just when I try to pay Priya Lee and it has changed since last time?

Overall, I don't fully disagree with the "Update Contact" button, I just think the usage diverges from how it's phrased in the BIP.

When I see "Update Contact", that user story is that Stephen wants to pay Priya, so he types in her address and his wallet stores the offer, and then he pays the offer. But the BIP user story is more like, Stephen wants to pay Priya, so he types in her address and the wallet checks to figure out how to pay her, and whatever the wallet gets in response from the DNS servers we can reasonably trust because it has been signed by DNSSEC and if DNSSEC is broken then the internet is broken.

GBKS replied:
Good points, will update.

> Furthermore, I'm not sure how actionable it is -- you hit view changes, and then it shows you that it changed from one nasty looking lnobc1.... string to another nasty lnobc1... string.

Just knowing that something unexpectedly changed can be helpful, even if you can't interpret the change instantly and need to dig a little.

GBKS replied:
Added a line of text about this happening at the point of initiating a payment, and will revisit in #1083.

sbddesign commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,285 @@
+
+### How it works
+
+Lightning addresses look like traditional emails. Under the hood, a wallet that uses a lightning address reformats it to a URL. It then looks for information about a [LNURL](https://github.com/lnurl/luds) web-server at that location (more on that [here](https://github.com/lnurl/luds/blob/luds/16.md)). That server handles the communication with the lightning node and passes messages back and forth ...
```
```suggestion
Lightning addresses look like traditional emails. Under the hood, a wallet that uses a lightning address reformats it to a URL. It then looks for information about a [LNURL](https://github.com/lnurl/luds) web-server at that location (more on that [here](https://github.com/lnurl/luds/blob/luds/16.md)). That server handles the communication with the lightning node and passes messages back and forth. This allows for complex interactions (like authentication and withdrawals), but also requires trusting the web server.
```

While I absolutely love the DNS payment instructions, to be completely fair, it also requires technical overhead to implement. So I don't think it's accurate or fair to characterize Lightning address as "this adds technical overhead".

sbddesign reviewed (commented):
Very good page. I'm glad to see this topic covered.

I went through and provided some thoughts: small grammar nits, one visual nit, and some technical commentary. Happy to jump on a call and discuss sometime if you like.

mouxdesign commented:
Adding in some birds eye view ideas as well. More on the overarching structure of the content, less about the copy itself.

Reading through the content it forms 4 larger chunks:
- Introduction
- DNS Payment instructions
- Lightning addresses 
- Paynyms

- Paynyms is also mentioned as a smaller paragraph right at the end, it would be good to mention it at the beginning in the sentence "On this page, we focus primarily on the [DNS Payment Instructions proposal with some information about lightning addresses _**and Paynyms.**_"
- The introduction has 1) The basic idea and 2) Address format and basics. With the section on security and privacy inbetween. It feels as if the two basics sections could be together here as they both lean into the format topic.
- Lightning addresses: There is the headings address format + how it works in DNS section. The address format is also mentioned right at the beginning of the lightning section so could be a possible heading there. 
- Sharing screen design: A question that comes up here is how does a user know that one address is their private address and the other one their public address. If there is a way to make this more obvious on the design to prevent them for mistakenly sharing the wrong one.

GBKS commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,285 @@
+
+Every approach requires at least one intermediary that is being contacted to serve the payment information.
+
+This brings up important questions around privacy and security. Intermediaries can potentially serve up different payment information than a user specified. Or they can track requests and metadata to build profiles of users and user behavior. If these intermediate hops are not fully controlled by the user, the address should be considered custodial (or at least semi-custodial).
```
Maybe the problem is that we are so used to the term "custodial" being used for the funds themselves - who controls the key - and now we use it for addresses with very different context. Not sure how to resolve this other than using a different term. What do you think?

Generally, I find the paragraph clear as a whole, even if the specific term could be off.

GBKS replied:
I clarified the text a bit further, describing what is meant by custodial here. Hope that addresses your feedback.

GBKS commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,285 @@
+
+### How it works
+
+Users create DNS entries with payment information, which can be one or more addresses of different formats, combined to a [BIP 21 URIs](https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki). Since address reuse is best avoided for privacy reasons, the included addresses should only be static payment codes (like [silent payment codes](https://gist.github.com/RubenSomsen/c43b79517e7cb701e ...
```
I think I read that, and then someone else said that DNS propagation is not speedy and trustworthy enough to make it super reliable. No particular thoughts on this, other than recommending to stick with BOLT12 and silent payments.

🤷

GBKS commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,285 @@
+
+#### Setup
+
+Hosting human-readable addresses carries legal and financial responsibility, so many non-custodial wallets are unlikely to offer this feature. However, they should still make it easy to copy payment information for setting up addresses on domains owned by the user or a third party of their choice.
```
I was thinking more about the service needing to run servers and databases to handle the user data, probably some form of login, etc. Not like Bitcoin Core, for example, where it's a standalone piece of software you run locally and don't get any other service. So the wallet provider would need to be a company and everything that comes with it. Does that make sense?

sbddesign replied:
Sure, it's something we consider are considering with [TwelveCash](https://twelve.cash/) (how do we best handle users editing data). However, I disagree with the statement "so many non-custodial wallets are unlikely to offer this feature".

You can facilitate *pushing the user's offer to your DNS records* without taking *custody of the user's funds*. That aspect is pretty trivial. In fact, it's arguable that many non-custodial wallets would want to use this as a selling point (that you can have a cool username and be self-custodial).

sbddesign replied:
> So the wallet provider would need to be a company and everything that comes with it.

Being a company does not equal being custodial

GBKS replied:
I am not thinking of custody of funds, just custody of the address.

Let's imagine you create your address with some anon service online. After an initial period where everything works well, they start occasionally swapping out your actual payment info with someone else's. Or maybe they sell the domain or service to someone else, who starts messing things up. Or maybe they beef up prices and threaten to just relay your payment info to your own if you don't pay up. Or they get hacked and someone replaces the info right before your monthly paycheck comes in and you don't find out until later. If the service simply stops working and addresses break, then it's not so bad. But other scenarios could also happen.

So I think there is a responsibility that comes with owning the relay of this one piece of information to another piece of information. This is not the case when you share your BOLT12 or silent payments addresses directly. Would Sparrow Wallet or a similar project want this responsibility? Without being a company and having the required legal protection?

In the progressive security sense, there are maybe two realities. In a "Nostr zap via Lightning Address" world of tiny payments and small amounts, maybe it's fine to just be more relaxed. But with meaningful funds and use, then it's a different story. `Mo' money, mo' security`, if you will.

GBKS replied:
Also, TwelveCash looks cool.

GBKS replied:
I lightened the text a bit more on this responsibility aspect, hope that addresses the feedback.

GBKS commented on assets/images/guide/how-it-works/human-readable-addresses/setup.png:
Thanks for taking a close look at those. There's a separate issue for revising these setup screens [here](https://github.com/BitcoinDesign/Guide/issues/1083), since the ones in this PR are more conceptual, as the text indicates. I kept them conceptual, because it was just too much to handle in one PR and I just needed a bit more time and brain focus to push them further. Your feedback will definitely flow into that iteration, but I'd like prefer not to address them right here in this PR if that's OK.

I don't understand your second rationale. Just because someone else could leak my information or somehow find it by brute-force scraping guessing, does not mean I should not be careful about what I post publicly. Privacy practices are often about reducing the chance that others see something.

GBKS commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,285 @@
+
+### Trust
+
+Setting this up requires a user to own a domain, and trust the domain provider and DNS system to truthfully communicate their payment information. If the user does not own the domain but uses a third-party service instead, then another intermediary is in play.
```
I think we're reading this paragraph very differently.

> "setting this up requires a user to own a domain"

If you want to set this up yourself, you need to own a domain.

> If the user does not own the domain but uses a third-party service instead

If you don't have a domain, you can use a third-party.

> then another intermediary is in play

That third-party is an extra intermediary that you don't have if you own the domain. That changes the trust assumption (from "registrar + DNS" to "third-party + registrar + DNS").

Looking at Nostr, there is a difference getting my NIP-05 address via [nostrplebs.com](https://nostrplebs.com/) or through an address I have registered myself.

So I think we're mostly on the same page here, but just reading/interpreting things differently. What do you think?

GBKS commented:
I think I addressed all feedback. There are some points around the UI visuals that I will address in a follow-up in #1083. Please take a look and let me know how it looks know. Now that conference season is over, I'd like to wrap this PR up at a good pace. Thanks in advance.

GBKS commented:
Strike [now supports](https://strike.me/blog/bolt12-offers/) BOLT12 and DNS-based human readable addresses, along with Zeus, Phoenix and others (full list on [bolt12.org](https://bolt12.org)). Posting this here, due to the [editorial question](https://github.com/BitcoinDesign/Guide/issues/1101) about proposals with little or no adoption. More adoption over time speaks in favor of this content.

GBKS commented:
[Selfie Records](https://selfie-records.com) extends BIP353 with additional data types. Another sign of adoption. Nonetheless by Synonym, so it sounds like @BitcoinErrorLog may have embraced it as well.

---

> I'd go further and say [Proton Wallet](https://github.com/ProtonWallet)'s "Bitcoin via email" is the biggest step thus far in the direction of human-readable addresses and should be added to this page as an example.

@yashrajd, this seems like an internal solution by a private company, and not an open standard, right? Not sure that really belongs on the page, unless there's more to their way of doing things.

BitcoinErrorLog commented:
fwiw we are not currently pro- or anti-Selfie Records. The project was made by a motivated team member and we thought it was a good way to express capabilities DNS always had, to frame the design space clearly. All namespaces that want to enforce uniqueness (squatting) require a central authority, but that does not negate that some people find it useful and are willing to take that risk to solve what they consider to be problems, despite its censorability and difficulty for individuals to self-manage.

Our actual love letter to decentralized identity, routing & payment coordination is PKARR _(and Paykit, but R&D for that is paused til January)_ https://github.com/pubky/pkarr -- All of the public keys and endpoint records people want to be putting in DNS or relays or such should be put into Mainline DHT with PKARR. Because it is objectively the most decentralized and resilient network on the internet afaik. Projects like Iroh and Web5 are already using this setup, but Bitcoiners typically prefer bitcoin/LN-dedicated solutions. PKARR doesnt need blockchains or bitcoin keys.

We still arent sure whether/how we might support Selfies in our own products, but we have a bookmark on approaching the "nickname" problem ideally/holistically someday if it seems needed.

Sorry for shilling, but we wouldn't be building these things if we didn't think they were the best solutions :)

sbddesign reviewed (commented):
Basically LGTM. However, in the time since this PR was drafted, the convention switched form `b12` to `lno`. Update the `lno` in all the images and it has my stamp.

ConorOkus commented:
The self-hosted/DIY path will add a lot of friction, and I don't think many people own personal domains. It may be something more suited to people running businesses/merchants/freelancers etc. 

Feedback from the Phoneix implementation suggests users are having a few issues https://x.com/methobit/status/1811496354549760022

ConorOkus commented:
What's the latest thinking on naming conventions for the BIP353 address and  BOLT 12 offers?

GBKS commented:
I just pushed an update to this page with lots of changes to address various points of feedback from the past months.

- Revision of the address setup flow (which I thought I'd do in #1083, but here it is)
- Added notes about user-facing naming (probs will end up in discussion, as naming tends to do)
- Lots of small copy tweaks and additions
- Added a table of contents at the top
- Revision of diagrams for clarity
- Change b12 to lno to match the updated convention (per comment by @sbddesign) 
- Updated backlink from the private key management intro page
- Updated from master

Would be amazing to get some fresh reviews on this. There may also have been other changes to the proposal or learnings from real-world implementations that should be considered. Let me know if you are aware of anything.

@ConorOkus I added a paragraph about naming, and some of the mock-ups also include specific choices of labels. I'll also point to a [recent update I posted](https://gbks.substack.com/p/112-a-rose) with some thoughts around naming, mostly to point out that naming is a highly contextual thing and cannot just be "solved globally" like 2 + 2. Please take a look and let me know what you think (about this page, not necessarily the linked post).

Happy 2025 everyone 🕺💃🪩

yashrajd commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,339 @@
+
+Below is a sample flow, in which a user can define their unique username, which is then combined with the wallet providers domain to form their human-readable address.
+
+{% include image-gallery.html pages = page.images_setup %}
```
My sense is: This user flow only sets up the username. If the payment information is automatically added by the application, it should be shown to the user as confirmation or it should be mentioned. If the user is expected to configure it further, that should be indicated as well... what do you think @GBKS ?

GBKS replied:
Goal of this design was to provide a simple happy path, but allow for customization if a user is interested in that (but it is not expected of the user).

I don't think we need to show the payment info. Would a user understand what they are looking at when they see a BOLT 12 offer? Could they even catch an error if there is one? Instead, maybe an additional info screen with some explainers could be helpful?

swedishfrenchpress replied:
I think the onboarding screens are fine, in the example below it shows the options for primary and backup which are clear to me.

yashrajd commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,339 @@
+
+Address settings allow users to adjust the payment information associated with the human-readable address.
+
+{% include image-gallery.html pages = page.images_settings %}
```
This is great! 💯

yashrajd commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,339 @@
+
+The additional configuration is important, to account for the diversity of features that different wallets offer. There are likely to be situations where the user wants to bundle payment information from several wallets. The best candidates for address types to use (due to their re-usability) are BOLT12 offers and silent payment addresses. Those are not broadly established, so a user may rely on  ...
+
+#### Sharing
```
This is great advice. It also fits very well with contacts, we should mention it here and link to the guidance on Contacts!

yashrajd commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,339 @@
+
+## Lightning Address
+
+Compared to DNS Payment Instructions, [Lightning Addresses](http://lightningaddress.com) are focused on the lightning network only. They bring another intermediary into play, and allow for much more dynamic functionality.
```
> They bring another intermediary into play, and allow for much more dynamic functionality.

- which another intermediary? (just for my knowledge)
- what more dynamic functionality? (perhaps we can mention it?)

GBKS replied:
Both things are explained in the following paragraph.

yashrajd commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,339 @@
+   height = 159
+%}
+
+In a similar vein, [UMA](https://www.uma.me) is based on LNURL and Lightning Address, with modified functionality and the *"$<span class="-green">alice</span>@<span class="-blue">domain.com</span>"* address format. The *$* prefix was chosen to represent both fiat and crypto currencies.
```
I oppose mentioning UMA since it amounts to promoting it. It supports shitcoinery, compliance garbage, is connected to threat actors, and is generally bad IMO. Perhaps somebody can articulate the criticism better...

yashrajd commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,339 @@
+
+## [Paynyms](http://paynym.is)
+
+This approach relies on a single directory provider, which maps a human readable name with a [BIP-47](https://github.com/bitcoin/bips/blob/master/bip-0047.mediawiki) payment code. Having a single provider allows for the omission of the global part, in this case *"my.paynym.is/<span class="-green">username</span>"* and users can simply be referred to by their usernames. The directory code is not o ...
```
Perhaps it is unwise to mention it if it's open-source? If we're mentioning non-open-source stuff, might be a good idea to mention Proton's "bitcoin via email" as well.

yashrajd reviewed (commented):
Lots of great additions over previous version of this PR, this is turning out to be a fantastic page with great explanations, mockups and advice! 

This is not great protocol for privacy or self-sovereignty but pretty good from a usability angle, the evergreen tradeoff.

GBKS commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,339 @@
+The benefits to the user experience include:
+
+- Familiarity: A format that feels more human and familiar to users
+- Ease of sharing: A format that is easier to remember and recall and share in a social setting
```
How about `in-person`?

GBKS commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,339 @@
+- Familiarity: A format that feels more human and familiar to users
+- Ease of sharing: A format that is easier to remember and recall and share in a social setting
+
+**Important:** Human-readable addresses should only be used for direct communication. Whenever payment information is communicated digitally (copy & paste, links, QR codes, messaging, etc), cryptographic formats should be used for the best security and privacy.
```
I disagree. Two points to this. First, wallets can copy the underlying payment info and not the human-readable address, so there are things wallets can do. And second, a lot of best practices are up to individuals to implement and cannot (or should not) be enforced. If some people don't follow them, that does not make them irrelevant (eating healthy is still a good idea if someone does not).

yashrajd replied:
> I disagree. Two points to this. First, wallets can copy the underlying payment info and not the human-readable address, so there are things wallets can do. 

Since these addresses would be easy to remember, recall and share (as above line mentions), the user is likely to simply type it into chat/email/social media.

> And second, a lot of best practices are up to individuals to implement and cannot (or should not) be enforced. If some people don't follow them, that does not make them irrelevant (eating healthy is still a good idea if someone does not).

Fair point, but as you mention these addresses are **meant** to be shared widely and assumed to be public, so such a suggestion counter-signals the very idea. The fact is that this is not a security/privacy-friendly payment method. 

Thus, instead of this 'important note' for the audience of the guide, a better approach might be to recommend that a warning/disclaimer/note be shown to the end-user regarding the privacy/security lacunas involved (you already mention these on this page later)

yashrajd replied:
Added the above suggestion as a Tip in the Security & privacy section [here](https://github.com/BitcoinDesign/Guide/pull/1082#discussion_r1911313625)

GBKS replied:
Going to resolve this comment, since you moved the suggestion to another comment, and I'll respond there.

GBKS commented:
Thank you, @yashrajd. I accepted most of your tweaks and left a few comments. Whether to include UMA and Paynyms in this page is a good question. I'd be happy to remove them as I don't think they add very much value. What do others think?

swedishfrenchpress commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,339 @@
+
+### How it works
+
+Users create DNS entries with payment information, which can be one or more addresses of different formats, combined to a [BIP 21 URI](https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki). Since address reuse is best avoided for privacy reasons, the included addresses should be reusable (like [silent payments]({{ '/guide/how-it-works/silent-payments/' | relative_url }}) and [BOLT12](ht ...
```
How does a user create a DNS entry? What does that mean? Do they do that in their bitcoin wallet? Or is that something they do with the on their website domain via DNS settings? Some more detail would help me understand this.

GBKS replied:
This manual setup is something they would do with their domain name registrar. Typically looks something [like this](https://help.one.com/hc/en-us/articles/360000825478-How-do-I-create-a-TXT-record). It's something for the more technical crowed that is comfortable owning and managing domains.

swedishfrenchpress commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,339 @@
+Framing the correct use of the address should be done right away and as early in the user flow as possible. This allows user to make informed decisions as to how and when to use the address. The information should include:
+
+- The purpose of the address is for verbally sharing their bitcoin address in a social setting.
+- That the address is public, which carries a clear directional understanding about privacy.
```
I like the idea of making the comparison to an email address.

yashrajd commented on guide/how-it-works/human-readable-addresses.md:
```diff
@@ -0,0 +1,339 @@
+- Services can potentially serve different payment information than specified
+
+If the domain (website) where the payment information is stored, is not controlled by the user, we will refer to the address as "_managed_". To be clear, intermediaries cannot move funds at the user-provided address. They can only prevent payment information retrieval, or re-route funds by returning different payment information.
+
```
Providing this as an alternative to the "important note" earlier in the page, mentioned in [this comment thread](https://github.com/BitcoinDesign/Guide/pull/1082/files#r1911263474)

```suggestion
{% include tip/open.html color="blue" label="Tip" icon="info" %}

Inform the user that there is a privacy & security trade-off involved in setting up a managed address. If the application supports a user-provided domain, explain that this method minimises the risk and help them set it up with detailed instructions.

{% include tip/close.html %}

```

GBKS replied:
Hm, so I am not sure it is the wallet's responsibility to provide instructions for setting things up in third-party domain registrars. Maybe a link out to a website with more info could be an option? This might be good to have either way for users to learn a bit more about the feature before they decide whether to use it.

But the way I thought about this is that the user can skip setup of the Bubble Wallet address and just enter their own human-readable address in [their own profile in contacts](https://bitcoin.design/guide/daily-spending-wallet/contacts/#a-contact-card-for-the-wallet-owner). If this address is meant to be communicated verbally, then there's not a huge benefit of having it stored in your wallet (maybe so you don't forget it).

yashrajd replied:
Why not? That way they don't even have to support BIP-353 themselves, but if they do it serves as a cool feature. Since the TXT record on the domain have to contain all the payment info (bolt-12 and/or silent payments addresses etc.) potentially provided by the wallet using the BIP-21 format (see [Phoenix instructions to BTCSessions](https://x.com/phoenixwallet/status/1811455293886255602)), it seems prudent to have this. Having a third-party generate/provide these might be risky...what do you think?

GBKS replied:
My point still stands:

> Maybe a link out to a website with more info could be an option? This might be good to have either way for users to learn a bit more about the feature before they decide whether to use it.

ConorOkus commented:
Looking good @GBKS . I like rolling with bitcoin address. The address settings page is a little confusing to me though. We have Bubble address, Primary bitcoin address, Bubble wallet lightning address, Backup bitcoin address. I like [third party] bitcoin address (Bubble bitcoin address). Since we default to the wallet-provided BOLT 12 offer perhaps that can just be hidden, but in the event a user wants to change/edit that or add an sp address they can be referred to as "payment codes".

Maybe primary and secondary payment codes?

GBKS commented:
@ConorOkus I agree that the screen could become a little simpler. Take a look at the attached variation, which is a bit simpler and more like what a user would see the first time they enter this screen.

![image](https://github.com/user-attachments/assets/e55e5245-6038-401f-aa3a-a44eeffdc6bd)

It's a bit like when you come to the address tab when making an online purchase and need to define both a billing address and a shipping address.

Overall, I really think that we can (and should) avoid any additional terms (like payment code) with just the right tweaks to usability and copywriting. After all, if you include a new term, then you have to explain that one, too, and address it in all the other relevant screens.

GBKS commented:
> I can't think of many places other than hidden in settings where we would surface the "reusable payment codes"? Feels like we should only use the term "bitcoin address/bitcoin [third party] address" when referring to the actual human-readable name.

Do you mean just in the context of this imaginary wallet or the whole ecosystem?

> The second screen with the bottom sheet "choose a bitcoin address" is referring to the payment codes and not the human-readable name. Given it's editable, people can copy/paste, send these out of band etc we still need a name for them.

Yes, re-usable address, or just address. Nice and easy.

> I'm not sure how much explaining it would need though, traditional banks don't seem to bother explaining the differences between "the long card number", "account number", "sort code", "security code", "verification number" etc

Those are all different types of information. What we're talking about in the context of this page is different variations of the same type. Human-readable address, bitcoin address, re-usable address, are all things I can enter into the same input field to make a payment. They just have some different properties. My general proposal is that we call it all "address" and use "modifiers" to highlight the most relevant property in the context (re-usable, lightning, etc).

GBKS commented on assets/images/guide/how-it-works/human-readable-addresses/mapping.png:
Good point. I updated the diagram.

GBKS commented:
Just pushed another update. @ConorOkus and I had a call last week about naming. The gist is that where you end up with naming has a lot to do with your starting assumptions. Conor's approach is pretty forward-looking, using the assumption that human-readable addresses will be the default (and therefore will be called _bitcoin addresses_), _payment codes_ will be the term for the underlying re-usable address types (BOLT-12, BIP-47, BIP-352), and _onchain bitcoin address_ will stand for P2TR, etc.

It is pretty straightforward for us to individually come up with naming conventions that make sense to us. It's a lot harder and more squishy to reason about ecosystem-wide conventions, especially ones in the future. To navigate that in this PR, I created a special note that explains the naming rationale used in the mock-ups and states that wallets can come up with their own approaches. I hope that adds clarity, so a reader can understand the mock-ups for what they are meant to convey.

Let me know if that makes sense. I appreciate all the input and conversations.

sbddesign reviewed (approved):
<img width="548" alt="Screenshot 2025-01-23 at 22 41 05" src="https://github.com/user-attachments/assets/267e9671-47b1-421b-82b2-00a369c2e2ed" />
